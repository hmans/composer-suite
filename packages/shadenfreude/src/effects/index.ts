import { Color } from "three"
import {
  Add,
  Bitangent,
  Cross,
  Mul,
  Multiply,
  Normalize,
  Remap,
  Simplex3DNoise,
  Smoothstep,
  Step,
  Sub,
  Subtract,
  Tangent,
  VertexNormal,
  VertexPosition
} from "../nodes"
import { Value } from "../tree"

export const Dissolve = (
  visibility: Value<"float"> = 0.5,
  scale: Value<"float"> = 1,
  edgeThickness: Value<"float"> = 0.1,
  edgeColor: Value<"vec3"> = new Color(0, 10, 8)
) => {
  const noise = Remap(
    Simplex3DNoise(Multiply(VertexPosition, scale)),
    -1,
    Add(edgeThickness, 1),
    0,
    1
  )

  return {
    color: Multiply(
      edgeColor,
      Smoothstep(Subtract(visibility, edgeThickness), visibility, noise)
    ),

    alpha: Step(noise, visibility)
  }
}

export const UpdateVertexNormal = (
  modifier: (v: Value<"vec3">) => Value<"vec3">,
  offset = 0.001
) => {
  const tangent = Tangent(VertexNormal)
  const bitangent = Bitangent(VertexNormal, tangent)

  const neighbor1 = Add(VertexPosition, Mul(tangent, offset))
  const neighbor2 = Add(VertexPosition, Mul(bitangent, offset))

  const displacedNeighbor1 = modifier(neighbor1)
  const displacedNeighbor2 = modifier(neighbor2)

  const position = modifier(VertexPosition)
  const displacedTangent = Sub(displacedNeighbor1, position)
  const displacedBitangent = Sub(displacedNeighbor2, position)

  return Normalize(Cross(displacedTangent, displacedBitangent))
}
