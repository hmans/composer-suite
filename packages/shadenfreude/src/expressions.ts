import { glslRepresentation } from "./glslRepresentation"

export type Expression = {
  _: "Expression"
  values: any[]
  render: () => string
}

const zip = (a: TemplateStringsArray, b: any[]) => a.map((k, i) => [k, b[i]])

export const code = (
  strings: TemplateStringsArray,
  ...values: any[]
): Expression => ({
  _: "Expression",

  values,

  render: () =>
    zip(strings, values.map(glslRepresentation))
      .flat()
      .join("")
})

export function isExpression(v: any): v is Expression {
  return v && v._ === "Expression"
}
