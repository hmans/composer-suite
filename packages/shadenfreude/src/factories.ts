import { GLSLType, ShaderNode, Variable } from "./types"
import { float, variablesToNodes } from "./variables"

export type ShaderNodeTemplate<T extends GLSLType> = Partial<ShaderNode<T>>

export function node<T extends GLSLType>(template: ShaderNodeTemplate<T>) {
  /* Create node from template */
  const node: ShaderNode<T> = {
    _shaderNode: true,
    name: "Unnamed",
    uniforms: {},
    varyings: {},
    inputs: {},
    outputs: {},
    vertex: { header: "", body: "" },
    fragment: { header: "", body: "" },
    ...template
  }

  /* Register outputs */
  for (const [_, variable] of Object.entries(node.outputs)) {
    variablesToNodes.set(variable, node)
  }

  return node
}

export function plug(source: Variable) {
  return {
    into: (target: Variable) => {
      target.value = source
    }
  }
}
