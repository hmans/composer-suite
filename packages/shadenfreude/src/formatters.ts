import { Color, Vector2, Vector3, Vector4 } from "three"

export function formatInt(n: number) {
  return Math.round(n).toString()
}

export function formatFloat(n: number) {
  const s = n.toString()
  return s.includes(".") || s.includes("e") ? s : s + ".0"
}

export function formatVec2(a: number, b: number) {
  return `vec2(${formatValue(a)}, ${formatValue(b)})`
}

export function formatVec3(a: number, b: number, c: number) {
  return `vec3(${formatValue(a)}, ${formatValue(b)}, ${formatValue(c)})`
}

export function formatVec4(a: number, b: number, c: number, d: number) {
  return `vec4(${formatValue(a)}, ${formatValue(b)}, ${formatValue(
    c
  )}, ${formatValue(d)})`
}

export function formatColor(color: Color) {
  return formatVec3(color.r, color.g, color.b)
}

export function formatValue(value: any): string {
  if (typeof value === "string") {
    return value
  } else if (typeof value === "number") {
    return formatFloat(value)
  } else if (value instanceof Color) {
    return formatColor(value)
  } else if (value instanceof Vector2) {
    return formatVec2(value.x, value.y)
  } else if (value instanceof Vector3) {
    return formatVec3(value.x, value.y, value.z)
  } else if (value instanceof Vector4) {
    return formatVec4(value.x, value.y, value.z, value.w)
  } else {
    throw new Error(`Could not render value to a GLSL representation: ${value}`)
  }
}
