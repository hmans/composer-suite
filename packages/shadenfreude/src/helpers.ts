import { Color, Matrix3, Matrix4, Vector2, Vector3, Vector4 } from "three"
import { GLSLType, isShaderNode, isVariable, Value } from "./types"

export function glslType(value: Value): GLSLType {
  if (isVariable(value)) {
    return value.type
  } else if (isShaderNode(value)) {
    return glslType(value.outputs.value)
  } else if (typeof value === "number") {
    return "float"
  } else if (typeof value === "boolean") {
    return "bool"
  } else if (value instanceof Color) {
    return "vec3"
  } else if (value instanceof Vector2) {
    return "vec2"
  } else if (value instanceof Vector3) {
    return "vec3"
  } else if (value instanceof Vector4) {
    return "vec4"
  } else if (value instanceof Matrix3) {
    return "mat3"
  } else if (value instanceof Matrix4) {
    return "mat4"
  } else {
    throw new Error(`Could not find a GLSL type for: ${value}`)
  }
}
