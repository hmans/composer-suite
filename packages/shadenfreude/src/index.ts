export * from "./compilers"
export * from "./factories"
export * from "./helpers"
export * from "./nodes"
export * from "./types"
