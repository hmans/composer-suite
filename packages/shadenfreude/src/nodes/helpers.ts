import { GLSLType, Variable } from "../variables"

export const Pipe = <T extends Variable<GLSLType>>(
  v: T,
  ...ops: ((v: T) => T)[]
) => ops.reduce((acc, op) => op(acc), v)
