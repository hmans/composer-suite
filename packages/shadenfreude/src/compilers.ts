import { RenderCallback } from "@react-three/fiber"
import { IUniform } from "three"
import { variablesToNodes } from "./factories"
import { formatValue } from "./formatters"
import { isVariable, ShaderNode, Value, Variable } from "./types"

type Program = "vertex" | "fragment"

export function compileVariableValue(value: Value): string {
  return isVariable(value) ? value.name : formatValue(value)
}

export function compileVariable(variable: Variable) {
  const valueString =
    variable.qualifier !== "uniform" && variable.value !== undefined
      ? ` = ${compileVariableValue(variable.value)}`
      : ""

  return `${variable.qualifier ?? ""} ${variable.type} ${
    variable.name
  }${valueString};`
}

function nodeTitle(node: ShaderNode) {
  return `/** Node: ${node.name} **/`
}

function dependencies(node: ShaderNode, deps = new Array<ShaderNode>()) {
  for (const variable of Object.values(node.inputs)) {
    if (isVariable(variable.value)) {
      /* get dependency */
      const dependency = variablesToNodes.get(variable.value)!
      if (!dependency) throw new Error("Dependency not found")

      /* If we haven't seen this dependency yet, invoke the callback */
      if (!deps.includes(dependency)) {
        dependencies(dependency, deps)
      }
    }
  }

  deps.push(node)
  return deps
}

function compileHeader(node: ShaderNode, program: Program) {
  const header = (n: ShaderNode) => `
    ${nodeTitle(n)}

    ${Object.values(n.uniforms)
      .map((variable) => compileVariable({ ...variable, qualifier: "uniform" }))
      .join("\n")}

    ${Object.values(n.varyings)
      .map((variable) => compileVariable({ ...variable, qualifier: "varying" }))
      .join("\n")}

    ${n[program].header ?? ""}
  `

  return dependencies(node)
    .map(header)
    .join("\n\n\n")
}

function compileBody(node: ShaderNode, program: Program) {
  return dependencies(node)
    .map(
      (node) => `
        ${nodeTitle(node)}

        ${Object.entries(node.outputs)
          .map(([_, variable]) =>
            compileVariable({ ...variable, value: undefined })
          )
          .join("")}

        {
          /* Varying References */
          ${Object.entries(node.varyings)
            .map(([name, variable]) =>
              compileVariable({ ...variable, name, value: variable.name })
            )
            .join("")}

          /* Inputs */
          ${Object.entries(node.inputs)
            .map(([name, variable]) => compileVariable({ ...variable, name }))
            .join("")}

          /* Outputs */
          ${Object.entries(node.outputs)
            .map(([name, variable]) => compileVariable({ ...variable, name }))
            .join("")}

          /* Code */
          ${node[program].body ?? ""}

          /* Update globals */
          ${Object.entries(node.outputs)
            .map(([name, variable]) => `${variable.name} = ${name};`)
            .join("\n")}
          ${
            program === "vertex"
              ? Object.entries(node.varyings)
                  .map(([name, variable]) => `${variable.name} = ${name};`)
                  .join("")
              : ""
          }
          }
      `
    )
    .join("\n\n\n")
}

export function compileShader(root: ShaderNode) {
  /* Build uniforms */
  const uniforms = {} as Record<string, IUniform>

  dependencies(root).forEach((node) => {
    Object.entries(node.uniforms).forEach(([name, variable]) => {
      uniforms[variable.name] = variable as IUniform
    })
  })

  /* Build callbacks */
  const callbacks = dependencies(root).map((node) => node.update)

  const update: RenderCallback = (...args) => {
    callbacks.forEach((callback) => callback?.(...args))
  }

  /* Build vertex shader program */
  const vertexShader = `
    /*** VERTEX SHADER ***/

    ${compileHeader(root, "vertex")}

    void main() {
      ${compileBody(root, "vertex")}
    }`

  /* Build fragment shader program */
  const fragmentShader = `
    /*** FRAGMENT SHADER ***/

    ${compileHeader(root, "fragment")}

    void main() {
      ${compileBody(root, "fragment")}
    }`

  return {
    vertexShader,
    fragmentShader,
    uniforms,
    update
  }
}
