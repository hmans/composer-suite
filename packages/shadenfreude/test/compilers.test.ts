import { Vector3 } from "three"
import { node, Variable, variable } from "../src"
import { compileShader, compileVariable } from "../src/compilers"
import "./helpers"

describe("compileShader", () => {
  it("should compile a shader", () => {
    const TestNode = (input: { offset?: Variable<"vec3"> } = {}) =>
      node({
        name: "Root",

        inputs: {
          offset: variable("vec3", input.offset || new Vector3(1, 2, 3))
        },

        vertex: {
          body: "csm_Position += offset;"
        }
      })

    const shader = compileShader(TestNode())

    expect(shader).toMatchSnapshot()
  })
})

describe("compileVariable", () => {
  it("compiles the given variable into its GLSL representation", () => {
    const v = variable("float", 1)
    expect(compileVariable(v)).toBeGLSL(`float ${v.name} = 1.0;`)
  })

  it("uses string values verbatim (assuming they are a reference to a variable)", () => {
    const v = variable("float", "foo")
    expect(compileVariable(v)).toBeGLSL(`float ${v.name} = foo;`)
  })

  it("renders non-scalar values using their GLSL representation", () => {
    const v = variable("vec3", new Vector3(1, 2, 3))
    expect(compileVariable(v)).toBeGLSL(`vec3 ${v.name} = vec3(1.0, 2.0, 3.0);`)
  })

  it("resolves variable references", () => {
    const source = variable("float", 1)
    const v = variable("float", source)
    expect(compileVariable(v)).toBeGLSL(`float ${v.name} = ${source.name};`)
  })
})
