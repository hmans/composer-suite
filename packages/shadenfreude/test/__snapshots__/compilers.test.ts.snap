// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`compileShader compiles a node into a shader 1`] = `
Object {
  "fragmentShader": "
    /*** FRAGMENT SHADER ***/

    
    /** Node: Root **/

    

    

    
  

    void main() {
      
        /** Node: Root **/

        

        {
          /* Varying References */
          

          /* Inputs */
           vec3 offset = vec3(1.0, 2.0, 3.0);

          /* Outputs */
          

          /* Code */
          

          /* Update globals */
          

          
        }
      
    }",
  "uniforms": Object {},
  "update": [Function],
  "vertexShader": "
    /*** VERTEX SHADER ***/

    
    /** Node: Root **/

    

    

    
  

    void main() {
      
        /** Node: Root **/

        

        {
          /* Varying References */
          

          /* Inputs */
           vec3 offset = vec3(1.0, 2.0, 3.0);

          /* Outputs */
          

          /* Code */
          csm_Position += offset;

          /* Update globals */
          

          
        }
      
    }",
}
`;

exports[`compileShader resolves dependencies to other nodes 1`] = `
Object {
  "fragmentShader": "
    /*** FRAGMENT SHADER ***/

    
    /** Node: Constant Color Value **/

    

    

    
  



    /** Node: Constant Vector3 Value **/

    

    

    
  



    /** Node: CSM Master Node **/

    

    

    
  

    void main() {
      
        /** Node: Constant Color Value **/

         vec3 node_1_constant_color_value_output_vec3_value;

        {
          /* Varying References */
          

          /* Inputs */
          

          /* Outputs */
           vec3 value = vec3(1.0, 0.4117647058823529, 0.7058823529411765);

          /* Code */
          

          /* Update globals */
          node_1_constant_color_value_output_vec3_value = value;

          
        }
      



        /** Node: Constant Vector3 Value **/

         vec3 node_2_constant_vector3_value_output_vec3_value;
 float node_2_constant_vector3_value_output_float_x;
 float node_2_constant_vector3_value_output_float_y;
 float node_2_constant_vector3_value_output_float_z;

        {
          /* Varying References */
          

          /* Inputs */
          

          /* Outputs */
           vec3 value = vec3(1.0, 2.0, 3.0);
 float x = value.x;
 float y = value.y;
 float z = value.z;

          /* Code */
          

          /* Update globals */
          node_2_constant_vector3_value_output_vec3_value = value;
node_2_constant_vector3_value_output_float_x = x;
node_2_constant_vector3_value_output_float_y = y;
node_2_constant_vector3_value_output_float_z = z;

          
        }
      



        /** Node: CSM Master Node **/

        

        {
          /* Varying References */
          

          /* Inputs */
           vec3 diffuseColor = node_1_constant_color_value_output_vec3_value;
 vec3 emissiveColor;
 vec3 position = node_2_constant_vector3_value_output_vec3_value;
 vec3 normal;

          /* Outputs */
          

          /* Code */
          
        csm_DiffuseColor.rgb = diffuseColor;
        
        
      

          /* Update globals */
          

          
        }
      
    }",
  "uniforms": Object {},
  "update": [Function],
  "vertexShader": "
    /*** VERTEX SHADER ***/

    
    /** Node: Constant Color Value **/

    

    

    
  



    /** Node: Constant Vector3 Value **/

    

    

    
  



    /** Node: CSM Master Node **/

    

    

    
  

    void main() {
      
        /** Node: Constant Color Value **/

         vec3 node_1_constant_color_value_output_vec3_value;

        {
          /* Varying References */
          

          /* Inputs */
          

          /* Outputs */
           vec3 value = vec3(1.0, 0.4117647058823529, 0.7058823529411765);

          /* Code */
          

          /* Update globals */
          node_1_constant_color_value_output_vec3_value = value;

          
        }
      



        /** Node: Constant Vector3 Value **/

         vec3 node_2_constant_vector3_value_output_vec3_value;
 float node_2_constant_vector3_value_output_float_x;
 float node_2_constant_vector3_value_output_float_y;
 float node_2_constant_vector3_value_output_float_z;

        {
          /* Varying References */
          

          /* Inputs */
          

          /* Outputs */
           vec3 value = vec3(1.0, 2.0, 3.0);
 float x = value.x;
 float y = value.y;
 float z = value.z;

          /* Code */
          

          /* Update globals */
          node_2_constant_vector3_value_output_vec3_value = value;
node_2_constant_vector3_value_output_float_x = x;
node_2_constant_vector3_value_output_float_y = y;
node_2_constant_vector3_value_output_float_z = z;

          
        }
      



        /** Node: CSM Master Node **/

        

        {
          /* Varying References */
          

          /* Inputs */
           vec3 diffuseColor = node_1_constant_color_value_output_vec3_value;
 vec3 emissiveColor;
 vec3 position = node_2_constant_vector3_value_output_vec3_value;
 vec3 normal;

          /* Outputs */
          

          /* Code */
          
        
        csm_Position = position;
        
      

          /* Update globals */
          

          
        }
      
    }",
}
`;
